In [1]:
import numpy as np
import pandas as pd
from scipy.io import arff

### Loading data and extracting only necessary colums

- age   	Age
- bp	    Blood Pressure
- sg	    Specific Gravity
- al	    Albumin
- su	    Sugar
- bgr	    Blood Glucose Random
- bu	    Blood Urea
- sod	    Sodium
- pot	    Potassium
- hemo	    Hemoglobin
- pcv	    Packed Cell Volume
- wbcc	    White Blood Cell Count
- rbcc	    Red Blood Cell Count
- class 	Class

In [14]:
data = arff.loadarff('../datasets/ckd/chronic_kidney_disease_full.arff')
df = pd.DataFrame(data[0])
df = df[["age", "bp", "sg", "al", "su", "bgr", "bu", "sod", "pot", "hemo", "pcv", "wbcc", "rbcc", "class"]]
df.head(10)

,age,bp,sg,al,su,bgr,bu,sod,pot,hemo,pcv,wbcc,rbcc,class
0,48.0,80.0,b'1.020',b'1',b'0',121.0,36.0,NaN,NaN,15.4,44.0,7800.0,5.2,b'ckd'
1,7.0,50.0,b'1.020',b'4',b'0',NaN,18.0,NaN,NaN,11.3,38.0,6000.0,NaN,b'ckd'
2,62.0,80.0,b'1.010',b'2',b'3',423.0,53.0,NaN,NaN,9.6,31.0,7500.0,NaN,b'ckd'
3,48.0,70.0,b'1.005',b'4',b'0',117.0,56.0,111.0,2.5,11.2,32.0,6700.0,3.9,b'ckd'
4,51.0,80.0,b'1.010',b'2',b'0',106.0,26.0,NaN,NaN,11.6,35.0,7300.0,4.6,b'ckd'
5,60.0,90.0,b'1.015',b'3',b'0',74.0,25.0,142.0,3.2,12.2,39.0,7800.0,4.4,b'ckd'
6,68.0,70.0,b'1.010',b'0',b'0',100.0,54.0,104.0,4.0,12.4,36.0,NaN,NaN,b'ckd'
7,24.0,NaN,b'1.015',b'2',b'4',410.0,31.0,NaN,NaN,12.4,44.0,6900.0,5.0,b'ckd'
8,52.0,100.0,b'1.015',b'3',b'0',138.0,60.0,NaN,NaN,10.8,33.0,9600.0,4.0,b'ckd'
9,53.0,90.0,b'1.020',b'2',b'0',70.0,107.0,114.0,3.7,9.5,29.0,12100.0,3.7,b'ckd'


### Converting Columns To Show Desired Values

In [27]:
# Converting the "hemo" column to g/l from g/dl by dividing by 10.
df["hemo"] = df["hemo"] / 10
df.head(10)

,age,bp,sg,al,su,bgr,bu,sod,pot,hemo,pcv,wbcc,rbcc,class
0,48.0,80.0,b'1.020',b'1',b'0',121.0,36.0,NaN,NaN,1.54,44.0,7800.0,5.2,b'ckd'
1,7.0,50.0,b'1.020',b'4',b'0',NaN,18.0,NaN,NaN,1.13,38.0,6000.0,NaN,b'ckd'
2,62.0,80.0,b'1.010',b'2',b'3',423.0,53.0,NaN,NaN,0.96,31.0,7500.0,NaN,b'ckd'
3,48.0,70.0,b'1.005',b'4',b'0',117.0,56.0,111.0,2.5,1.12,32.0,6700.0,3.9,b'ckd'
4,51.0,80.0,b'1.010',b'2',b'0',106.0,26.0,NaN,NaN,1.16,35.0,7300.0,4.6,b'ckd'
5,60.0,90.0,b'1.015',b'3',b'0',74.0,25.0,142.0,3.2,1.22,39.0,7800.0,4.4,b'ckd'
6,68.0,70.0,b'1.010',b'0',b'0',100.0,54.0,104.0,4.0,1.24,36.0,NaN,NaN,b'ckd'
7,24.0,NaN,b'1.015',b'2',b'4',410.0,31.0,NaN,NaN,1.24,44.0,6900.0,5.0,b'ckd'
8,52.0,100.0,b'1.015',b'3',b'0',138.0,60.0,NaN,NaN,1.08,33.0,9600.0,4.0,b'ckd'
9,53.0,90.0,b'1.020',b'2',b'0',70.0,107.0,114.0,3.7,0.95,29.0,12100.0,3.7,b'ckd'


In [33]:
#show all types of class column
df["class"].unique()

#change b"ckd" = a, b"notckd" = c
df["class"] = df["class"].replace({b"ckd": "a", b"notckd": "c"})

#checking the first 10 rows of the dataframe after the changes
df.head(10)

,age,bp,sg,al,su,bgr,bu,sod,pot,hemo,pcv,wbcc,rbcc,class
0,48.0,80.0,b'1.020',b'1',b'0',121.0,36.0,NaN,NaN,1.54,44.0,7800.0,5.2,a
1,7.0,50.0,b'1.020',b'4',b'0',NaN,18.0,NaN,NaN,1.13,38.0,6000.0,NaN,a
2,62.0,80.0,b'1.010',b'2',b'3',423.0,53.0,NaN,NaN,0.96,31.0,7500.0,NaN,a
3,48.0,70.0,b'1.005',b'4',b'0',117.0,56.0,111.0,2.5,1.12,32.0,6700.0,3.9,a
4,51.0,80.0,b'1.010',b'2',b'0',106.0,26.0,NaN,NaN,1.16,35.0,7300.0,4.6,a
5,60.0,90.0,b'1.015',b'3',b'0',74.0,25.0,142.0,3.2,1.22,39.0,7800.0,4.4,a
6,68.0,70.0,b'1.010',b'0',b'0',100.0,54.0,104.0,4.0,1.24,36.0,NaN,NaN,a
7,24.0,NaN,b'1.015',b'2',b'4',410.0,31.0,NaN,NaN,1.24,44.0,6900.0,5.0,a
8,52.0,100.0,b'1.015',b'3',b'0',138.0,60.0,NaN,NaN,1.08,33.0,9600.0,4.0,a
9,53.0,90.0,b'1.020',b'2',b'0',70.0,107.0,114.0,3.7,0.95,29.0,12100.0,3.7,a


### Cleaning Missing Values


In [ ]:
# Count missing values per row
missing_values_count = df.isna().sum(axis=1)

# Dropping rows with 3 or more missing values
df_cleaned = df[missing_values_count < 3]

df_cleaned.head(10)

# show the number or rows left after dropping rows with 3 or more missing values
print(f"Number of rows left after dropping rows with 3 or more missing values: {len(df_cleaned)}")


Number of rows left after dropping rows with 3 or more missing values: 292


### Splitting Dataframes
Splitting the dataframes to seperate the affected and control individuals.


In [55]:
df_affected_group = df_cleaned[df_cleaned["class"] == "a"]
df_control_group = df_cleaned[df_cleaned["class"] == "c"]

display(df_affected_group.head(10))
print(f"Number of rows in the affected group: {len(df_affected_group)}")

display(df_control_group.head(10))
print(f"Number of rows in the control group: {len(df_control_group)}")

,age,bp,sg,al,su,bgr,bu,sod,pot,hemo,pcv,wbcc,rbcc,class
0,48.0,80.0,b'1.020',b'1',b'0',121.0,36.0,NaN,NaN,1.54,44.0,7800.0,5.2,a
3,48.0,70.0,b'1.005',b'4',b'0',117.0,56.0,111.0,2.5,1.12,32.0,6700.0,3.9,a
4,51.0,80.0,b'1.010',b'2',b'0',106.0,26.0,NaN,NaN,1.16,35.0,7300.0,4.6,a
5,60.0,90.0,b'1.015',b'3',b'0',74.0,25.0,142.0,3.2,1.22,39.0,7800.0,4.4,a
6,68.0,70.0,b'1.010',b'0',b'0',100.0,54.0,104.0,4.0,1.24,36.0,NaN,NaN,a
8,52.0,100.0,b'1.015',b'3',b'0',138.0,60.0,NaN,NaN,1.08,33.0,9600.0,4.0,a
9,53.0,90.0,b'1.020',b'2',b'0',70.0,107.0,114.0,3.7,0.95,29.0,12100.0,3.7,a
11,63.0,70.0,b'1.010',b'3',b'0',380.0,60.0,131.0,4.2,1.08,32.0,4500.0,3.8,a
12,68.0,70.0,b'1.015',b'3',b'1',208.0,72.0,138.0,5.8,0.97,28.0,12200.0,3.4,a
14,68.0,80.0,b'1.010',b'3',b'2',157.0,90.0,130.0,6.4,0.56,16.0,11000.0,2.6,a


Number of rows in the affected group: 148


,age,bp,sg,al,su,bgr,bu,sod,pot,hemo,pcv,wbcc,rbcc,class
250,40.0,80.0,b'1.025',b'0',b'0',140.0,10.0,135.0,5.0,1.50,48.0,10400.0,4.5,c
251,23.0,80.0,b'1.025',b'0',b'0',70.0,36.0,150.0,4.6,1.70,52.0,9800.0,5.0,c
252,45.0,80.0,b'1.025',b'0',b'0',82.0,49.0,147.0,4.4,1.59,46.0,9100.0,4.7,c
253,57.0,80.0,b'1.025',b'0',b'0',119.0,17.0,135.0,4.7,1.54,42.0,6200.0,6.2,c
254,51.0,60.0,b'1.025',b'0',b'0',99.0,38.0,135.0,3.7,1.30,49.0,8300.0,5.2,c
255,34.0,80.0,b'1.025',b'0',b'0',121.0,27.0,144.0,3.9,1.36,52.0,9200.0,6.3,c
256,60.0,80.0,b'1.025',b'0',b'0',131.0,10.0,146.0,5.0,1.45,41.0,10700.0,5.1,c
257,38.0,60.0,b'1.020',b'0',b'0',91.0,36.0,135.0,3.7,1.40,46.0,9100.0,5.8,c
258,42.0,80.0,b'1.020',b'0',b'0',98.0,20.0,140.0,3.5,1.39,44.0,8400.0,5.5,c
259,35.0,80.0,b'1.020',b'0',b'0',104.0,31.0,135.0,5.0,1.61,45.0,4300.0,5.2,c


Number of rows in the control group: 144


### Statistics

In [58]:
print("Descriptive statistics for the affected group:")
display(df_affected_group.describe(include='all'))

print("Descriptive statistics for the control group:")
display(df_control_group.describe(include='all'))

Descriptive statistics for the affected group:


,age,bp,sg,al,su,bgr,bu,sod,pot,hemo,pcv,wbcc,rbcc,class
count,145.000000,146.000000,148,148,148,140.000000,146.000000,129.000000,128.000000,147.000000,146.000000,127.000000,120.000000,148
unique,NaN,NaN,6,7,7,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1
top,NaN,NaN,b'1.010',b'0',b'0',NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,a
freq,NaN,NaN,53,32,85,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,148
mean,57.606897,81.027397,NaN,NaN,NaN,178.542857,82.058219,133.003876,4.735156,1.045714,31.986301,9256.692913,3.911667,NaN
std,13.997171,16.177014,NaN,NaN,NaN,91.867387,59.887985,13.625358,3.869965,0.222437,7.262550,3638.388588,0.866189,NaN
min,6.000000,50.000000,NaN,NaN,NaN,22.000000,1.500000,4.500000,2.500000,0.310000,9.000000,2600.000000,2.100000,NaN
25%,50.000000,70.000000,NaN,NaN,NaN,107.000000,39.000000,131.000000,3.900000,0.910000,28.000000,6900.000000,3.400000,NaN
50%,60.000000,80.000000,NaN,NaN,NaN,153.000000,62.500000,136.000000,4.300000,1.050000,32.000000,9100.000000,3.900000,NaN
75%,66.000000,90.000000,NaN,NaN,NaN,238.250000,107.000000,139.000000,4.900000,1.190000,36.000000,10800.000000,4.400000,NaN


Descriptive statistics for the control group:


,age,bp,sg,al,su,bgr,bu,sod,pot,hemo,pcv,wbcc,rbcc,class
count,143.000000,142.000000,144,144,144,140.000000,140.000000,141.000000,141.000000,142.000000,143.000000,141.000000,141.000000,144
unique,NaN,NaN,3,2,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1
top,NaN,NaN,b'1.020',b'0',b'0',NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,c
freq,NaN,NaN,70,139,139,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,144
mean,46.622378,71.267606,NaN,NaN,NaN,107.471429,33.042857,141.780142,4.338298,1.518944,46.377622,7682.978723,5.383688,NaN
std,15.843337,8.576659,NaN,NaN,NaN,18.588866,11.411764,4.766089,0.592170,0.127717,4.150011,1834.158330,0.593973,NaN
min,12.000000,60.000000,NaN,NaN,NaN,70.000000,10.000000,135.000000,3.300000,1.300000,40.000000,4300.000000,4.500000,NaN
25%,34.000000,60.000000,NaN,NaN,NaN,93.000000,24.000000,138.000000,3.700000,1.410000,43.000000,6300.000000,4.900000,NaN
50%,46.000000,70.000000,NaN,NaN,NaN,107.000000,33.000000,141.000000,4.500000,1.500000,46.000000,7500.000000,5.300000,NaN
75%,58.000000,80.000000,NaN,NaN,NaN,123.250000,44.000000,146.000000,4.900000,1.617500,50.000000,9200.000000,5.900000,NaN


There are some values that are physically impossible to have, for example potassium in the affected group is recorded at maximum value of 47, which is physically impossible. The minimum value of sodium in the affected group is 4.5, which is way too low. These are certainly data errors, a human can't survive with those values. There are some more values that are slightly outside the norms like bloodpressure, but sodium and potassium were the most dramatic outliers.

We would handle them, by dropping the whole rows that have dramatic statistical errors. 

### Correlation Matrix

for aleksput